int32 curl1
int32 hype1
int32 curl2
int32 hype2
int32 curl3
int32 hype3
int32 curl4
int32 hype4
